app/role_test.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/mattermost/mattermost-server/v5/utils/slices"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetRolesByNames(t *testing.T) {
+	testPermissionInheritance(t, func(t *testing.T, th *TestHelper, testData permissionInheritanceTestData) {
+		actualRoles, err := th.App.GetRolesByNames([]string{testData.channelRole.Name})
+		require.Nil(t, err)
+		require.Len(t, actualRoles, 1)
+
+		actualRole := actualRoles[0]
+		require.NotNil(t, actualRole)
+		require.Equal(t, testData.channelRole.Name, actualRole.Name)
+
+		require.Equal(t, testData.shouldHavePermission, slices.IncludesString(actualRole.Permissions, testData.permission.Id))
+	})
+}
+
+func TestGetRoleByName(t *testing.T) {
+	testPermissionInheritance(t, func(t *testing.T, th *TestHelper, testData permissionInheritanceTestData) {
+		actualRole, err := th.App.GetRoleByName(testData.channelRole.Name)
+		require.Nil(t, err)
+		require.NotNil(t, actualRole)
+		require.Equal(t, testData.channelRole.Name, actualRole.Name)
+		require.Equal(t, testData.shouldHavePermission, slices.IncludesString(actualRole.Permissions, testData.permission.Id))
+	})
+}
+
+func testPermissionInheritance(t *testing.T, testCallback func(t *testing.T, th *TestHelper, testData permissionInheritanceTestData)) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	th.App.SetLicense(model.NewTestLicense(""))
+	th.App.SetPhase2PermissionsMigrationStatus(true)
+
+	permissionsDefault := []string{
+		model.PERMISSION_MANAGE_CHANNEL_ROLES.Id,
+		model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS.Id,
+	}
+
+	// Defer resetting the system scheme permissions
+	systemSchemeRoles, err := th.App.GetRolesByNames([]string{
+		model.CHANNEL_GUEST_ROLE_ID,
+		model.CHANNEL_USER_ROLE_ID,
+		model.CHANNEL_ADMIN_ROLE_ID,
+	})
+	require.Nil(t, err)
+	require.Len(t, systemSchemeRoles, 3)
+
+	// defer resetting the system role permissions
+	for _, systemRole := range systemSchemeRoles {
+		defer th.App.PatchRole(systemRole, &model.RolePatch{
+			Permissions: &systemRole.Permissions,
+		})
+	}
+
+	// Make a channel scheme, clear its permissions
+	channelScheme, err := th.App.CreateScheme(&model.Scheme{
+		Name:        model.NewId(),
+		DisplayName: model.NewId(),
+		Scope:       model.SCHEME_SCOPE_CHANNEL,
+	})
+	require.Nil(t, err)
+	defer th.App.DeleteScheme(channelScheme.Id)
+
+	team := th.CreateTeam()
+	defer th.App.PermanentDeleteTeamId(team.Id)
+
+	// Make a channel, set its scheme
+	channel := th.CreateChannel(team)
+	channel.SchemeId = &channelScheme.Id
+	channel, err = th.App.UpdateChannelScheme(channel)
+	require.Nil(t, err)
+
+	// See tests/channel-role-has-permission.md for an explanation of the below matrix.
+	truthTable := [][]bool{
+		{true, true, true, true},
+		{true, true, false, false},
+		{true, false, true, true},
+		{true, false, false, true},
+		{false, true, true, false},
+		{false, true, false, false},
+		{false, false, true, false},
+		{false, false, false, false},
+	}
```

Comment:
I know theres a readme about this truth table in a different file but do you think it would still be beneficial to add a comment for the column names here just so we dont have to go back and forth

Reply:
Added the columns.